game: Add initial C++ implementation of game code

alayander opened the pull request:
This implementation includes a terminal_display.cpp to emulate the output of the game, which will be replaced by the hologram fan. It just exists now in order to see output and test the game code.

alayander commented:
Opened in draft since there are a lot of open items left. These include:

- General refactoring of code
- Adding in comments
- Adding more common APIs for Game class to abstract to functions
- Cleaning up logic

alayander commented:
Will likely be adding a commit to this PR to include some change to CI stuff/clang-format since the tabbing seems unideal and the max char per length can likely be adjusted

BrianShTsoi commented on game/cpp/production/src/dino.cpp:
```diff
@@ -0,0 +1,53 @@
+
+int Dino::get_y_position() { return this->y_position; }
+
+void Dino::duck() {
+	if (this->state == Dino_State::JUMPING) {
+		this->jump();
+	} else {
+		this->state = Dino_State::DUCKING;
+		toggle_step();
+	}
+}
+
+void Dino::neutral() {
+	if (this->state == Dino_State::JUMPING) {
+		this->jump();
+	} else {
+		this->state = Dino_State::RUNNING;
+		toggle_step();
+	}
+}
+
+// TODO
+// clean up jump logic
+void Dino::jump() {
+	this->state = Dino_State::JUMPING;
+	this->y_position = JUMP_HEIGHTS[jump_index];
+	jump_index++;
+	if (jump_index == 45) {
+		jump_index = 0;
+		this->state = Dino_State::RUNNING;
+	}
+}
+
+void Dino::die() { this->state = Dino_State::DEAD; }
```
nit: The OCD in me thinks that we can probably combine these into a `try_set_state(Dino_State state)` method, then extract the jump logic to a separate step/function. This avoids some code duplication and also `jump(...)` being used in both methods in `Game` and methods in `Dino`.

alayander replied:
Resolved

BrianShTsoi commented on game/cpp/production/src/game.cpp:
```diff
@@ -0,0 +1,173 @@
+const std::array<std::array<int, DINO_WIDTH>, DINO_HEIGHT> &
+Game::fetch_dino_sprite() {
+	switch (this->dino.get_state()) {
+	case Dino_State::RUNNING:
+		switch (this->dino.get_step()) {
+		case Step_State::LEFT:
+			return DINO_LEFT_SPRITE;
+		case Step_State::RIGHT:
+			return DINO_RIGHT_SPRITE;
+		}
+	case Dino_State::DUCKING:
+		switch (dino.get_step()) {
+		case Step_State::LEFT:
+			return DINO_LEFT_DUCK_SPRITE;
+		case Step_State::RIGHT:
+			return DINO_RIGHT_DUCK_SPRITE;
+		}
```
``` Suggestion
	case Dino_State::RUNNING:
		return this->dino.get_step() == Step_State::LEFT ? DINO_LEFT_SPRITE : DINO_RIGHT_SPRITE;
	case Dino_State::DUCKING:
		return this->dino.get_step() == Step_State::LEFT ? DINO_LEFT_DUCK_SPRITE : DINO_RIGHT_DUCK_SPRITE;
```
nit: Just a personal preference, for the inner switch statements ternary operator looks nicer

alayander replied:
Resolved

BrianShTsoi commented on game/cpp/production/src/game.cpp:
```diff
@@ -0,0 +1,173 @@
+	bool collision = false;
+
+	draw_dino_sprite();
+
+	std::vector<Obstacle> obstacles = obs_manager.fetch_obstacles();
+	for (auto obs_itr = obstacles.begin(); obs_itr != obstacles.end();
+		 ++obs_itr) {
+		switch (obs_itr->get_type()) {
+		case Obstacle_Type::SMALL_CACTUS:
+			if (obs_itr == obstacles.begin()) {
+				collision = draw_small_cactus_sprite(obs_itr->get_location());
+			} else {
+				draw_small_cactus_sprite(obs_itr->get_location());
+			}
+			break;
+		case Obstacle_Type::LARGE_CACTUS:
+			if (obs_itr == obstacles.begin()) {
+				collision = draw_large_cactus_sprite(obs_itr->get_location());
+			} else {
+				draw_large_cactus_sprite(obs_itr->get_location());
+			}
+			break;
+		case Obstacle_Type::LOW_BIRD:
+		case Obstacle_Type::HIGH_BIRD:
+			if (obs_itr == obstacles.begin()) {
+				collision = draw_bird_sprite(obs_itr->get_location());
+			} else {
+				draw_bird_sprite(obs_itr->get_location());
+			}
+		}
+	}
+
+	// TODO
+	// make dead dino sprite not overwrite obstacle
+	if (collision) {
+		dino.die();
+		draw_dino_sprite();
+	}
+	return collision;
+}
```
nit: I am conflicted about whether I prefer returning the `colllision` bool (both in this function and the `draw_*_sprite(...)` functions), or maybe `Game` should have an attribute to store collision.
I think the functions names don't make it obvious that it is detecting collision, but having it as a return value possibly makes it more clear that this is something that the caller should handle. Oh well idk.

It also makes me slightly uneasy that the functions are drawing and detecting collision at the same time, but that's probably too much idealist thinking.

alayander replied:
Haha yeah this is honestly just me spaghetti code throwing this in at 4am to make it work

alayander replied:
Resolved: naming of `draw_*_sprite` -> `draw_*_with_collision`

The reason drawing and detecting collisions occur at the same time is because both can occur through the same nested for loop structure. If this were to be split into two separate functions would be less efficient and the collision work would have to apply transformations already happening to enable drawing with offset of the obstacle.

BrianShTsoi commented on game/cpp/production/src/game.cpp:
```diff
@@ -0,0 +1,173 @@
+		 ++obs_itr) {
+		switch (obs_itr->get_type()) {
+		case Obstacle_Type::SMALL_CACTUS:
+			if (obs_itr == obstacles.begin()) {
+				collision = draw_small_cactus_sprite(obs_itr->get_location());
+			} else {
+				draw_small_cactus_sprite(obs_itr->get_location());
+			}
```
```suggestion
			collision |= draw_small_cactus_sprite(obs_itr->get_location());
```

alayander replied:
Resolved

BrianShTsoi commented on game/cpp/production/src/main.cpp:
```diff
@@ -0,0 +1,63 @@
+std::atomic<char> curr_input{'\0'};
+bool quit = false;
+
+void input_thread_func() {
+	while (!quit) {
+		char key = read_key();
+		curr_input.store(key);
+	}
+}
```
I realize there is an issue with ducking, and it was due to the limitations on how I try to emulate user input with `read(...)`.

Basically when we try to press `s` to duck for an amount of time longer than a quick press. The dino will "debounce" after a quick duck, before going back to ducking. In a place as dangerous as the desert, this could mean death for the dino.

This is due to the key long press delay, a normal keyboard behaviour where when we press on a key for a long time, it has a slightly delay after the first character, before shooting out the character continuously. So from the program perspective, we get "s" then nothing for a little while then "ssssssssss...".

There might be ways to address this, I played around with the `read(...)` delay which could be increased to avoid the issue but then the key becomes way less responsive. I think we should probably just ignore this and move on. Obviously Arduino won't have this issue.

alayander replied:
Yup observed this too was going to make a general comment on this once PR flipped from draft to non-draft about how it won't matter since we will likely just poll a GPIO state/variable "live"

BrianShTsoi reviewed (commented):
This is looking really good. If I have time I'll try to make an ESP32 version to make sure it compiles (The internet has mixed accounts on whether Arduino works with STL/vectors/iterators, if not we'll live with C style arrays).

IMO this is good to land once the jump logic and obstacle logic is fixed. It's all pretty readable so I think comments can minimal, and I don't mind if we don't refactor some stuff, we can live with some duplication of code and it's all abstracted away anyways.

I put down some small nits, but feel free to ignore. I think we can prioritize starting ECE integration anyways, and we can always come back to refactor later.

alayander commented:
Yeah will address a lot of the nits, mentioned in a comment but quite a bit is just code thrown in late night to get a working state for the draft.

As for the nano esp32 compatibility I did some light research, so could be totally off. But was with the understanding that the ESP-based Arduino boards have compatibility/implementations with STL/vectors/etc. Obviously would need to be tested out, with some modifications (ie. Having the vector call `reserve()` to avoid/minimize reallocations and fragmentation - in out game flow there is maximum number of obstacles that will exists in the game at a given point relative to the `MAX_X ` and `OBSTACLE_COOLDOWN`.)

BrianShTsoi commented:
Just did a quick attempt at compiling with Nano ESP32 (only in the IDE, I didn't have a Nano ESP32 with me to try and flash it). It seems to compile fine (after slight modifications to get rid of the std library stuff obviously), while Arduino Uno would complain it doesn't recognize `<array>`. So yea ESP32 is probably fine. We just need to hope it runs properly too, tho can't really know for sure until we have fan display working

alayander committed:
ci: Update clang-format 100 max col size, switch case tabbing

Signed-off-by: Alexander Lay <alexanderqlay@gmail.com>

alayander committed:
game: Add initial C++ implementation of game code

This implementation includes a terminal_display.cpp to emulate the output of the game, which will be replaced by the hologram fan

Signed-off-by: Alexander Lay <alexanderqlay@gmail.com>

BrianShTsoi commented on game/cpp/production/src/dino.cpp:
```diff
@@ -0,0 +1,66 @@
+
+int Dino::get_y_position() { return this->y_position; }
+
+void Dino::try_set_input_state(Dino_State input_state) {
+	if (this->state == Dino_State::JUMPING) {
+		this->jump();
+	} else {
+		switch (input_state) {
+			case Dino_State::DUCKING:
+				this->duck();
+				break;
+			case Dino_State::JUMPING:
+				this->jump();
+				break;
+			default:
+				this->neutral();
+		}
+	}
+}
```
```suggestion
void Dino::try_set_input_state(Dino_State input_state) {
	if (this->state == Dino_State::JUMPING) {
		this->jump();
		return;
	} 
	switch (input_state) {
		case Dino_State::DUCKING:
			this->duck();
			break;
		case Dino_State::JUMPING:
			this->jump();
			break;
		default:
			this->neutral();
	}
}
```

BrianShTsoi commented on game/cpp/production/src/game.cpp:
```diff
@@ -0,0 +1,167 @@
+		case Dino_State::RUNNING:
+			return this->dino.get_step() == Step_State::LEFT ? DINO_LEFT_SPRITE : DINO_RIGHT_SPRITE;
+		case Dino_State::DUCKING:
+			return this->dino.get_step() == Step_State::LEFT ? DINO_LEFT_DUCK_SPRITE
+															 : DINO_RIGHT_DUCK_SPRITE;
```
Weird formatting?

alayander replied:
Exceeds 100 col max, so clang-format does the auto alignment

BrianShTsoi replied:
Oh tabs are 8-character wide on github that's why it looks weird

alayander commented:
The plan is for this PR to add initial implementation that works with the terminal display

A separate PR will be made to implement changes for compatibility to run on ESP32

BrianShTsoi reviewed (approved):
LGTM
